torch/csrc/jit/codegen/cuda/runtime/warp.cu

```diff
@@ -0,0 +1,62 @@
+namespace warp {
+
+const int WARP_SIZE = 32;
+
+template <
+    bool SINGLE_WARP,
+    typename T,
+    typename Func,
+    typename _dim3ti,
+    typename _dim3bd>
+__device__ void warpReduceTIDX(
+    T& out,
+    const T& inp_val,
+    Func reduction_op,
+    const _dim3ti& thread_idx,
+    const _dim3bd& block_dim,
+    T* shared_mem,
+    bool read_write_pred,
+    T init_val) {
+  // Assume input padded to multiples of a warp
+  T reduce_val = init_val;
+
+  // Do warp reduction
+  if (read_write_pred) {
+    reduce_val = inp_val;
+  }
+
+  // Reduce within each warp
+  for (int i = 16; i >= 1; i /= 2) {
+    reduction_op(reduce_val, __shfl_xor_sync(0xffffffff, reduce_val, i, 32));
+  }
+
+  // Reduce across warp if needed
+  // Load value to shared mem
+  if (!SINGLE_WARP) {
+    bool is_warp_head = thread_idx.x % 32 == 0;
+    unsigned int warp_idx = thread_idx.x / 32;
+    unsigned int reduction_size = block_dim.x;
+    unsigned int num_of_warps = reduction_size / 32;
+
+    block_sync::sync();
+
+    if (read_write_pred && is_warp_head) {
+      shared_mem[warp_idx] = reduce_val;
+    }
+
+    block_sync::sync();
+
+    if (is_warp_head && warp_idx == 0) {
+      // Reduce to each warp
+      for (int warp_it = 1; warp_it < num_of_warps; warp_it++) {
```

Comment:
If I read this correctly, this is executed only by thread 0. In the worst case, where `num_of_warps` is 32, parallelizing this serial execution might be beneficial. How about doing the final reduction also as a shuffle-based warp reduction?

Reply:
Updated thanks!